ocr.sh

```diff
@@ -1,5 +1,145 @@
-# OCR workflow to be run on ocrd_controller
-# (syntax of `ocrd process`)
-tesserocr-recognize -P segmentation_level region -P model frak2021 -I OCR-D-IMG -O OCR-D-OCR
-# we need ALTO in the end
-fileformat-transform -P from-to "page alto" -P script-args "--no-check-border --dummy-word" -I OCR-D-OCR -O FULLTEXT
+#!/bin/bash
+# OCR-D utils for ocr processing
+
+set -eu
+set -o pipefail
+
+TASK=$(basename $0)
+
+# initialize variables, create ord-d work directory and exit if something is missing
+# args:
+# 1. process ID
+# 2. task ID
+# 3. process dir path
+# 4. language
+# 5. script
+# 6. async (default true)
+# 7. workflow name (default preinstalled ocr-workflow-default.sh)
+# 8. images dir path under process dir(default images)
+# vars:
+# - CONTROLLER: host name and port of ocrd_controller for processing
+init () {    
+    logger -p user.info -t $TASK "ocr_init initialize variables and directory structure"
+    PROC_ID=$1
+    TASK_ID=$2
+    PROCDIR="$3"
+    LANGUAGE="$4"
+    SCRIPT="$5"
+    ASYNC=${6:-true}    
+    WORKFLOW="${7:-ocr-workflow-default.sh}"
+    PROCIMAGEDIR="${8:-images}"
+
+    logger -p user.notice -t $TASK "running with $* CONTROLLER=$CONTROLLER"
+    cd /data
+
+    if ! test -d "$PROCDIR"; then
+        logger -p user.error -t $TASK "invalid process directory '$PROCDIR'"
+        exit 2
+    fi
+    WORKFLOWFILE="$PROCDIR/ocr-workflow.sh"
+    if test -f "$WORKFLOWFILE"; then
+      WORKFLOW=$(realpath "$WORKFLOWFILE")
+    else
+      WORKFLOW=$(command -v "$WORKFLOW" || realpath "$WORKFLOW")
+        if ! test -f "$WORKFLOW"; then
+            logger -p user.error -t $TASK "invalid workflow '$WORKFLOW'"
+            exit 3
+        fi
+    fi
+    if test -z "$CONTROLLER" -o "$CONTROLLER" = "${CONTROLLER#*:}"; then
+        logger -p user.error -t $TASK "envvar CONTROLLER='$CONTROLLER' must contain host:port"
+        exit 4
+    fi
+    CONTROLLERHOST=${CONTROLLER%:*}
+    CONTROLLERPORT=${CONTROLLER#*:}
+
+    # copy the data from the process directory controlled by production
+    # to the transient directory controlled by the manager
+    # (currently the same share, but will be distinct volumes;
+    #  so the admin can decide to either mount distinct shares,
+    #  which means the images will have to be physically copied,
+    #  or the same share twice, which means zero-cost copying).
+    WORKDIR=ocr-d/"$PROCDIR" # will use other mount-point than /data soon
+    mkdir -p $(dirname "$WORKDIR")
+
+    pre_process_to_workdir
```

Comment:
IMO `pre_process_to_workdir` should be called from outside of `init` – to make the analogy with `post_process_to_ocrdir` clear.

Also these comments should be _inside_ the function, not at (or near) the call.

Reply:
Yes that makes much more sense!